Ks/loops lesson edits (#63)
Ada-Developers-Academy/ada-precourse-v2: 01-intro-to-python/05-loops.md

| for loop| A loop that _iterates_ or repeats over a sequence.  In python it is often used to iterate over a `list` data structure.  It is also commonly paired with the `range` [function](https://docs.python.org/3.3/library/stdtypes.html?highlight=range#ranges), which returns a list of numbers.  The `for` loop repeats n times, where n is the length of the sequence being looped over. | iterate | "To repeat the code five times we will use a for loop." |
| flag | A variable that is used to indicate that a condition has been met.  Usually a flag is a boolean variable and the value is changed to indicate that the goal or target (whatever the condition is) has been achieved. | signal | "The while loop will stop when the flag is changed." |

> ### Loop Examples

Author: Awesome, good to know. I'll get Prettier installed and reach out to folks if I have questions.

> ### Loop Examples

Author: Do you have any suggestions for using Prettier? It cleaned up some whitespace issues pretty well, but running it without changes from the defaults broke the challenges sections and didn't address table formatting like above (I made a manual pass so it's more readable like the tables in the challenges section).

Take a few moments to examine the different kinds of loops below, then guess what their output will be.

```python
# while loop
counter = 0
while counter < 10:
    counter = counter + 2
    print(f"The value of counter is {counter}")

# while loop with flag
flag_var = True
counter = 0
while flag_var:
    counter = counter + 3
    if(counter > 10):
        flag_var = False
    print(f"The value of counter is {counter}")

# for loop using ranges
for num in range(3):
    print(f"The value of num is {num}")

for num in range(2, 5):
    print(f"The value of num is {num}")

# for-each loop over a collection
for item in item_list:

>     print(item)

Reviewer: the later example uses number in number_list. Suggest maintaining the naming consistency.

We declare a while loop with the ‘while’ keyword, followed by the condition we want to evaluate, then a colon: 

```python
while <condition>:
    # Take actions inside the loop
```
With that in mind, let’s take another look at the first while loop we saw above.

```python
counter = 0
while counter < 10:
    counter = counter + 2
    print(f"The value of counter is {counter}")

# When run, this code will output:
# The value of counter is 2
# The value of counter is 4
# The value of counter is 6
# The value of counter is 8
# The value of counter is 10
```

In this example we’ve set up variable `counter` before the while loop.  The condition we check each time before entering the loop is whether the value of `counter` is less than 10.  Every pass through the loop we're increasing `counter` by 2, until it holds the value 10.  Once `counter` is holding the value 10, the expression `counter < 10` evaluates to false and we no longer enter the loop body.

What could it look like if we wanted to use a flag instead?

```python
flag_var = True
counter = 0

> while flag_var:

Author: I may not get to rewriting an example to include `break`, but I fully agree with you on the var name & will get that updated.

    counter = counter + 2
    if(counter >= 10):
        flag_var = False
    print(f"The value of counter is {counter}")

# When run, this code will output:
# The value of counter is 2
# The value of counter is 4
# The value of counter is 6
# The value of counter is 8
# The value of counter is 10
```

Here, we’ve declared the variables `counter` and `flag_var` before the loop.  The condition we check to enter our while loop is whether `flag_var` is set to true.  In the previous example we checked `counter`’s value directly in the while declaration, but with our flag, we instead check if `counter >= 10` inside our loop and use that to decide if `flag_var` should change to false. 

## `for` Loops

For loops let us loop over a fixed sequence, commonly a sequence of numbers, or items in a data structure.  This makes them our preferred choice when we know how many times we want to take an action. 

> For loops have a little more syntax than a while loop.  We declare a for loop using the `for` keyword, followed by a variable name (which we use inside the loop body to refer to the current value of the sequence we're iterating over), then the keyword `in`, followed by the sequence we are iterating over, and capped off with a colon:

Reviewer: - why is the syntax for for loops a little more involved. what does that get us? what are the trade offs?
- simplified the use of the iteration variable a little
- the first few examples use range rather than a concrete iterable (that students might recognize as being a sequence). suggest genericizing the description to an iterable expression rather than explicitly calling for a sequence.

```suggestion
For loops need a little bit more information than a while loop. While this makes the syntax of the loop itself a little more involved, on the other hand it means that all the information needed for the loop is collected in one place, making it a little more difficult to make mistakes!

We declare a for loop using the `for` keyword, followed by a variable name (which we use inside the loop body to refer to each iterated value), then the keyword `in`, followed by the iterable expression we are iterating over, and capped off with a colon:
```

Author: Thank you, I really like how you worded the "why".

```python

> for <variable_name> in <sequence_name>:

Reviewer: to match above suggestion

```suggestion
for <variable_name> in <iterable_expression>:
```

    # Do something with <variable_name>
```

> Say we want to print "Hello!" 3 times. When we know the exact number of times to take an action, we can use the `range` function to give us a list of numbers to iterate over.  `range` has some helpful options, if we pass it a single integer parameter, `range` will generate a sequence from 0 to the number you passed _exclusive_.  This means if you call `range(3)` it will create the sequence `[0, 1, 2]`. If we use this with a for loop, it would look something like this:  

Reviewer: - consistent voicing of we
- explicitly reinforce what we meant by saying the range is exclusive on the upper end
```suggestion
Say we want to print "Hello!" 3 times. When we know the exact number of times to take an action, we can use the `range` function to give us a list of numbers to iterate over.  `range` has some helpful options, if we pass it a single integer parameter, `range` will generate a sequence from 0 to the number we passed _exclusive_.  This means if we call `range(3)` it will create the sequence `[0, 1, 2]`, it _excludes_ 3. If we use this with a for loop, it would look something like this:  
```

```python

> for number in range(3):

Reviewer: The previous appearance used num rather than number (here and the next example). No strong opinion on which to use, but suggest keeping them the same.

    print("Hello!")
    print(f"The value of number is {number}")

# When run, this code will output:
# Hello!
# The value of number is 0
# Hello!
# The value of number is 1
# Hello!
# the value of number is 2
```

> If we need to perform work a specific number of times, but we don't want our count to start at 0, `range` still has us covered.  We can define the start of our number sequence by passing `range` two parameters, the number we want to start at (inclusive) and where we want to end (exclusive).  For example, if we call `range(2,5)` it will give us the number sequence `[2, 3, 4]`. Using this version of `range` in a for loop, we can do the following:

Reviewer: - here as well, explicitly point out what inclusive and exclusive mean
```suggestion
If we need to perform work a specific number of times, but we don't want our count to start at 0, `range` still has us covered.  We can define the start of our number sequence by passing `range` two parameters, the number we want to start at (inclusive) and where we want to end (exclusive).  For example, if we call `range(2,5)` it will give us the number sequence `[2, 3, 4]`, 2 is included but 5 is excluded. Using this version of `range` in a for loop, we can do the following:
```

```python
for number in range(2, 5):
    print(f"The value of number is {number}")

# When run, this code will output:
# The value of number is 2
# The value of number is 3
# the value of number is 4
```

> 

Reviewer: This would be a great spot to also bring up the skip argument for range, and show how to handle the original while loop example with a for loop.
